### PR TITLE
client/finality-grandpa: Reintegrate periodic neighbor packet worker

### PR DESCRIPTION
**Pull request by mxinden**
The `NeighborPacketWorker` within `client/finality-grandpa` does two
things:

1. It receives neighbor packets from components within
`client/finality-grandpa`, sends them down to the `GossipEngine` in
order for neighboring nodes to receive.

2. It periodically sends out the most recent neighbor packet to the
`GossipEngine`.

In order to send out packets it had a clone to a `GossipEgine` within
an atomic reference counter and a mutex. The `NeighborPacketWorker` was
then spawned onto its own asynchronous task.

Instead of running in its own task, this patch reintegrates the
`NeighborPacketWorker` into the main `client/finality-grandpa` task not
requiring the `NeighborPacketWorker` to own a clone of the
`GossipEngine`.

The greater picture

This is a tiny change within a greater refactoring. The overall goal is
to **simplify** how finality-grandpa interacts with the network and to
**reduce** the amount of **unbounded channels** within the logic.

Why no unbounded channels: Bounding channels is needed for backpressure
and proper scheduling. With unbounded channels there is no way of
telling the producer side to slow down for the consumer side to catch
up.  Rephrased, there is no way for the scheduler to know when to favour
the consumer task over the producer task on a crowded channel and the
other way round for an empty channel.

Reducing the amount of shared ownership simplifies the logic and enables
one to use async-await syntax-suggar, given that one does not need to
hold a lock across poll invocations. Using async-await enables one to
use bounded channels without complex logic.

### CODE REVIEW

File: client/finality-grandpa/src/communication/mod.rs [hunk 1]
```diff
@@ -134,7 +139,18 @@ pub(crate) struct NetworkBridge<B: BlockT, N: Network<B>> {
...
+	///
+	/// Packets send into this channel are processed by the `NeighborPacketWorker` and passed on to
+	/// the underlying `GossipEngine`.
 	neighbor_sender: periodic::NeighborPacketSender<B>,
+
+	/// `NeighborPacketWorker` processing packets send through the `NeighborPacketSender`.
+	//
+	// NetworkBridge is required to be clonable, thus one needs to be able to clone its children,
+	// thus one has to wrap neighor_packet_worker with an Arc Mutex.
+	neighbor_packet_worker: Arc<Mutex<periodic::NeighborPacketWorker<B>>>,
```
Comment (mxinden) [inline client/finality-grandpa/src/communication/mod.rs]: Yes, in order to reduce the amount of `Mutex`es, I introduce a new one here. This one is within the crate only and can be removed once `NetworkBridge` does not need to be clonable anymore.

File: client/finality-grandpa/src/communication/mod.rs [hunk 2]
```diff
@@ -134,7 +139,18 @@ pub(crate) struct NetworkBridge<B: BlockT, N: Network<B>> {
 	service: N,
 	gossip_engine: GossipEngine<B>,
 	validator: Arc<GossipValidator<B>>,
+
+	/// Sender side of the neighbor packet channel.
+	///
+	/// Packets send into this channel are processed by the `NeighborPacketWorker` and passed on to
+	/// the underlying `GossipEngine`.
 	neighbor_sender: periodic::NeighborPacketSender<B>,
```
Comment (mxinden) [inline client/finality-grandpa/src/communication/mod.rs]:
```suggestion
	neighbor_packet_sender: periodic::NeighborPacketSender<B>,
```

Happy to do this change, but rather as a follow up to reduce the noise within this pull request.

File: client/finality-grandpa/src/lib.rs [hunk 3]
```diff
@@ -889,9 +900,9 @@ pub fn run_grandpa<B, E, Block: BlockT, N, RA, SC, VR, X, Sp>(
 	Block::Hash: Ord,
 	B: Backend<Block> + 'static,
 	E: CallExecutor<Block> + Send + Sync + 'static,
-	N: NetworkT<Block> + Send + Sync + Clone + 'static,
+	N: NetworkT<Block> + Send + Sync + Clone + Unpin + 'static,
 	SC: SelectChain<Block> + 'static,
-	NumberFor<Block>: BlockNumberOps,
+	NumberFor<Block>: BlockNumberOps + Unpin,
```
Comment (tomaka) [inline client/finality-grandpa/src/lib.rs]: This is wrong

File: client/finality-grandpa/src/communication/periodic.rs [hunk 4]
```diff
@@ -56,56 +51,62 @@ impl<B: BlockT> NeighborPacketSender<B> {
...
-/// Does the work of sending neighbor packets, asynchronously.
-///
-/// It may rebroadcast the last neighbor packet periodically when no
-/// progress is made.
-pub(super) fn neighbor_packet_worker<B>(net: GossipEngine<B>) -> (
-	impl Future<Item = (), Error = ()> + Send + 'static,
-	NeighborPacketSender<B>,
-) where
-	B: BlockT,
+pub(super) struct NeighborPacketWorker<B: BlockT> {
```
Comment (rphmeier) [inline client/finality-grandpa/src/communication/periodic.rs]: could use a doc-string

File: client/finality-grandpa/src/communication/mod.rs [hunk 5]
```diff
@@ -391,6 +411,24 @@ impl<B: BlockT, N: Network<B>> NetworkBridge<B, N> {
 	}
 }
 
+impl<B: BlockT, N: Network<B>> Future03 for NetworkBridge<B, N>
+where
+	NumberFor<B>: Unpin,
```
Comment (tomaka) [inline client/finality-grandpa/src/communication/mod.rs]:
```suggestion
```
You probably have to add `impl<B, N> Unpin for NetworkBridge<B, N> {}` somewhere instead.

File: client/finality-grandpa/src/communication/periodic.rs [hunk 6]
```diff
@@ -56,56 +51,62 @@ impl<B: BlockT> NeighborPacketSender<B> {
...
+			last: None,
+			delay,
+			rx,
+		}, NeighborPacketSender(tx))
+	}
+}
+
+impl <B: BlockT> Stream for NeighborPacketWorker<B>
+where
+	NumberFor<B>: Unpin,
```
Comment (tomaka) [inline client/finality-grandpa/src/communication/periodic.rs]:
```suggestion
```

File: client/finality-grandpa/src/lib.rs [hunk 7]
```diff
@@ -551,10 +551,10 @@ pub fn run_grandpa_voter<B, E, Block: BlockT, N, RA, SC, VR, X, Sp>(
 	Block::Hash: Ord,
 	B: Backend<Block> + 'static,
 	E: CallExecutor<Block> + Send + Sync + 'static,
-	N: NetworkT<Block> + Send + Sync + Clone + 'static,
+	N: NetworkT<Block> + Send + Sync + Clone + Unpin + 'static,
 	SC: SelectChain<Block> + 'static,
 	VR: VotingRule<Block, Client<B, E, Block, RA>> + Clone + 'static,
-	NumberFor<Block>: BlockNumberOps,
+	NumberFor<Block>: BlockNumberOps + Unpin,
```
Comment (tomaka) [inline client/finality-grandpa/src/lib.rs]:
```suggestion
	NumberFor<Block>: BlockNumberOps,
```

File: client/finality-grandpa/src/lib.rs [hunk 8]
```diff
@@ -551,10 +551,10 @@ pub fn run_grandpa_voter<B, E, Block: BlockT, N, RA, SC, VR, X, Sp>(
 	Block::Hash: Ord,
 	B: Backend<Block> + 'static,
 	E: CallExecutor<Block> + Send + Sync + 'static,
-	N: NetworkT<Block> + Send + Sync + Clone + 'static,
+	N: NetworkT<Block> + Send + Sync + Clone + Unpin + 'static,
```
Comment (tomaka) [inline client/finality-grandpa/src/lib.rs]: Not sure about this `Unpin`

File: client/finality-grandpa/src/lib.rs [hunk 9]
```diff
@@ -650,13 +650,14 @@ struct VoterWork<B, E, Block: BlockT, N: NetworkT<Block>, RA, SC, VR> {
...
+	network: futures03::compat::Compat<NetworkBridge<Block, N>>,
 }
 
 impl<B, E, Block, N, RA, SC, VR> VoterWork<B, E, Block, N, RA, SC, VR>
 where
 	Block: BlockT,
- 	N: NetworkT<Block> + Sync,
-	NumberFor<Block>: BlockNumberOps,
+ 	N: NetworkT<Block> + Sync + Unpin,
+	NumberFor<Block>: BlockNumberOps + Unpin,
```
Comment (tomaka) [inline client/finality-grandpa/src/lib.rs]: Well, you get the idea

File: client/finality-grandpa/src/communication/mod.rs (hunk 5 above)
Comment (mxinden) [reply client/finality-grandpa/src/communication/mod.rs]: Thank you, thank you and thank you @tomaka :heart:

**Comment by mxinden** [comment]
@tomaka can you take another look?

File: client/finality-grandpa/src/communication/mod.rs [hunk 10]
```diff
@@ -134,9 +139,22 @@ pub(crate) struct NetworkBridge<B: BlockT, N: Network<B>> {
 	service: N,
 	gossip_engine: GossipEngine<B>,
 	validator: Arc<GossipValidator<B>>,
+
+	/// Sender side of the neighbor packet channel.
+	///
+	/// Packets send into this channel are processed by the `NeighborPacketWorker` and passed on to
```
Comment (andresilva) [inline client/finality-grandpa/src/communication/mod.rs]:
```suggestion
	/// Packets sent into this channel are processed by the `NeighborPacketWorker` and passed on to
```

File: client/finality-grandpa/src/communication/mod.rs [hunk 11]
```diff
@@ -134,9 +139,22 @@ pub(crate) struct NetworkBridge<B: BlockT, N: Network<B>> {
...
+	/// the underlying `GossipEngine`.
 	neighbor_sender: periodic::NeighborPacketSender<B>,
+
+	/// `NeighborPacketWorker` processing packets send through the `NeighborPacketSender`.
```
Comment (andresilva) [inline client/finality-grandpa/src/communication/mod.rs]:
```suggestion
	/// `NeighborPacketWorker` processing packets sent through the `NeighborPacketSender`.
```

**Pull request merged**